### ISSUE TITLE
Exception safety issues

### ISSUE THREAD

**Issue by osa1** [valderman/selda#39]
I was just reading the source code, and saw functions like this:

```haskell
-- | Perform the given computation over an SQLite database.
--   The database is guaranteed to be closed when the computation terminates.
withSQLite :: (MonadIO m, MonadMask m) => FilePath -> SeldaT m a -> m a
withSQLite file m = do
  conn <- sqliteOpen file
  runSeldaT m conn `finally` seldaClose conn
```

This functions is not thread-safe, if you get an asynchronous exception between `conn <- ...` and `runSeldaT ...` you leak the file descriptor.

Another example:

```haskell
-- | Run a Selda transformer. Backends should use this to implement their
--   @withX@ functions.
runSeldaT :: (MonadIO m, MonadMask m) => SeldaT m a -> SeldaConnection -> m a
runSeldaT m c = do
    liftIO $ takeMVar (connLock c)
    go `finally` liftIO (putMVar (connLock c) ())
  where
    ...
 ```

This is worse because you if you get an async exception between two lines all other threads get blocked indefinitely.

**Comment by valderman** [comment]
This issue is hopefully fixed by 17a49cff9b316b54eb821b930cca795646077860. I'm hesitant to close it without someone else taking a look at it, however.

**Comment by valderman** [comment]
We might also want to replace `MonadMask` for this with something based on `resourcet`, to help with cases like #38.

**Comment by valderman** [comment]
This should be fixed in 0.1.10.0. Please reopen if the issue pops up again.
